core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
     mStatistics = statistics;
     mUri = URI.create(mAlluxioHeader);
 
+    boolean masterAddIsSameAsDefault = checkMasterAddress();
+
     if (sInitialized) {
-      checkMasterAddress();
+      if (!masterAddIsSameAsDefault) {
+        throw new IOException(ExceptionMessage.DIFFERENT_MASTER_ADDRESS
+            .getMessage(mUri.getHost() + ":" + mUri.getPort(),
+                FileSystemContext.INSTANCE.getMasterAddress()));
+      }
+      setFileSystemAndContext();
       return;
     }
     synchronized (INIT_LOCK) {
```

Comment:
is it still meaningful to check static bool `INIT_LOCK` here, since we don't have filesystem handler and context static any more.